Client/WebFilters/DebouncingResourceDownloader.swift

```diff
@@ -0,0 +1,427 @@
+
+        for pattern in rule.include {
+          let urlPattern = URLPattern(validSchemes: .all)
+          let parseResult = urlPattern.parse(pattern: pattern)
+
+          switch parseResult {
+          case .success:
+            if urlPattern.isMatchingAllURLs {
+              // We put this query param and rule to our A bucket
+              // This seems to be an empty list for now
+              queryToRule[rule.param] = rule
+            } else if let etld1 = urlPattern.baseDomain {
+              // We put this etld and rule to our B bucket
+              var relevantPatterns = etldToRelevantPatterns[etld1] ?? []
+              relevantPatterns.append(pattern)
+              etldToRelevantPatterns[etld1] = relevantPatterns
+            } else {
+              // Everything else goes to our C bucket
+              // For the time being this set only encompases patterns without a host
+              otherRules.append(rule)
+            }
+          default:
+            log.error("Cannot parse pattern `\(pattern)`. Got status `\(parseResult)`")
+            continue
+          }
+        }
+
+        // Add or append the entry for each etld
+        for (etld1, relevantPatterns) in etldToRelevantPatterns {
+          var entries = etldToRule[etld1] ?? []
+          entries.append((rule, relevantPatterns))
+          etldToRule[etld1] = entries
+        }
+      }
+
+      self.etldToRule = etldToRule
+      self.queryToRule = queryToRule
+      self.otherRules = otherRules
+    }
+
+    /// Get redirect url recursively by continually applying the matcher rules to each url returned until we have no more redirects.
+    /// Only handles `http` and `https` schemes.
+    func redirectURLAll(from url: URL) -> URL? {
+      var redirectingURL = url
+      var result: URL?
+
+      while let nextURL = redirectURLOnce(from: redirectingURL) {
+        redirectingURL = nextURL
+        result = nextURL
+      }
+
+      return result
+    }
+
+    /// Get a possible redirect url for the given URL and the given matchers. Will only get the next redirect url.
+    /// Only handles `http` and `https` schemes.
+    ///
+    /// This code uses the patterns in the given matchers to determine if a redirect action is required.
+    /// If it is, a redirect url will be returned provided we can extract it from the url
+    ///
+    /// **Rules**
+    /// 1. For the given URL we *might* navigate too, pull out the `eTLD+1` and the query items
+    /// 2. If the eTLD+1 is in the `etldToRule` map (from step 2 above), see if the full URL matches the rule, and return the rule
+    /// (As an optimization, we only need to check the pattern the `eTLD+1` was extracted from and not the whole include list)
+    /// 3. If any of the query params (the keys) in the map from #4 above are in URL we might navigate to, apply the corresponding rule
+    /// 4. Apply all the rules from bucket C
+    func redirectURLOnce(from url: URL) -> URL? {
```

Comment:
nit : why `Once`?  I dont see any logic in the function that is checking to make sure something only happens once.  Am I missing something?  Otherwise this seems like an opportunity for confusion